File: src/idealync/cogs/role_selection.py
import logging

import discord
from discord.ext import commands

from idealync import Bot

from typing import Any;

logger = logging.getLogger(__name__)


class RoleSelect(discord.ui.Select[Any]):
    def __init__(self, bot: Bot) -> None:
        self.bot = bot
        options = [
            discord.SelectOption(label="Member", value="member", description="Members are also known as active contributors"),
            discord.SelectOption(label="Observer/Inactive", value="observer", description="Observer/Inactive are Members who are inactive for 1+ months and are unable to contribute for now"),
        ]
        super().__init__(
            placeholder="Choose your role",
            min_values=1,
            max_values=1,
            options=options,
        )

    async def callback(self, interaction: discord.Interaction) -> None:
        if interaction.guild is None:
            await interaction.response.send_message("this needs to be used in a server", ephemeral=True)
            return

        member = interaction.user
        if not isinstance(member, discord.Member):
            return

        target_role_id = (
            self.bot.config.member_role_id
            if self.values[0] == "member"
            else self.bot.config.observer_role_id
        )
        other_role_id = (
            self.bot.config.observer_role_id
            if self.values[0] == "member"
            else self.bot.config.member_role_id
        )

        target_role = interaction.guild.get_role(target_role_id)
        other_role = interaction.guild.get_role(other_role_id)

        if target_role is None:
            await interaction.response.send_message("That role is not available right now.", ephemeral=True)
            return

        if other_role is not None and other_role in member.roles:
            await member.remove_roles(other_role, reason="Switched role selection")

        if target_role not in member.roles:
            await member.add_roles(target_role, reason="Selected role from prompt")

        await interaction.response.send_message(
            f"You are now set as {target_role.name}.",
            ephemeral=True,
        )


class RoleSelection(commands.Cog):
    def __init__(self, bot: Bot) -> None:
        self.bot = bot
        self.role_message_id: int | None = None

    async def _find_existing_prompt(self, channel: discord.TextChannel) -> int | None:
        async for message in channel.history(limit=20):
            if message.author == self.bot.user and any(embed.title == "Choose your role" for embed in message.embeds):
                return message.id
        return None

    async def _ensure_role_prompt(self) -> None:
        channel = self.bot.get_channel(self.bot.config.role_channel_id)
        if not isinstance(channel, discord.TextChannel):
            logger.warning("Role channel %s is not a text channel.", self.bot.config.role_channel_id)
            return

        existing_message_id = await self._find_existing_prompt(channel)
        if existing_message_id is not None:
            self.role_message_id = existing_message_id
            return

        embed = discord.Embed(
            title="Choose your role",
            description="Use the dropdown below to switch between Member and Observer/Inactive.",
            color=discord.Color(0xF36647),
        )

        view = discord.ui.View()
        view.add_item(RoleSelect(self.bot))

        message = await channel.send(embed=embed, view=view)

        self.role_message_id = message.id
        logger.info("Posted role-selection prompt in channel %s", channel.id)

    @commands.Cog.listener()
    async def on_ready(self) -> None:
        await self._ensure_role_prompt()



async def setup(bot: Bot) -> None:
    await bot.add_cog(RoleSelection(bot))

File: src/main/kotlin/org/solync/idealync/cogs/RoleSelection.kt
package org.solync.idealync.cogs

import dev.kord.common.Color
import dev.kord.core.Kord
import dev.kord.core.behavior.channel.createMessage
import dev.kord.core.behavior.interaction.respondEphemeral
import dev.kord.core.behavior.interaction.response.respond
import dev.kord.core.entity.channel.TextChannel
import dev.kord.core.event.interaction.GuildSelectMenuInteractionCreateEvent
import dev.kord.core.on
import dev.kord.rest.builder.component.actionRow
import dev.kord.rest.builder.component.option
import dev.kord.rest.builder.message.embed
import org.solync.idealync.IdeaLyncModule
import org.solync.idealync.hasSelfEmbed
import org.solync.idealync.ideaLyncConfig

object RoleSelection : IdeaLyncModule {
    override suspend fun onRegister(kord: Kord) {
        kord.on<GuildSelectMenuInteractionCreateEvent> {
            if (interaction.componentId != "idealync:role_select") return@on

            val response = interaction.deferEphemeralResponse()
            val value = interaction.values[0]
            val addRoleId = if (value == "member") ideaLyncConfig.memberRoleId else ideaLyncConfig.observerRoleId
            val removeRoleId = if (value == "member") ideaLyncConfig.observerRoleId else ideaLyncConfig.memberRoleId

            val addRole = interaction.guild.getRoleOrNull(addRoleId)
            if (addRole == null) {
                interaction.respondEphemeral {
                    content = "That role is not available right now."
                }
                return@on
            }

            if (!interaction.user.roleIds.contains(addRoleId)) {
                interaction.user.addRole(addRoleId, "Selected role from prompt")
            }
            if (interaction.user.roleIds.contains(removeRoleId)) {
                interaction.user.removeRole(removeRoleId, "Switched role selection")
            }

            response.respond {
                content = "You are now set as ${addRole.name}."
            }
        }
    }

    override suspend fun onReady(kord: Kord) {
        ensureRolePrompt(kord)
    }

    private suspend fun ensureRolePrompt(kord: Kord) {
        val channel = requireNotNull(kord.getChannel(ideaLyncConfig.roleChannelId)) { "Role channel does not exist" }
        require(channel is TextChannel) { "Role channel is not a text channel." }

        if (channel.hasSelfEmbed("Choose your role")) {
            return
        }

        channel.createMessage {
            embed {
                title = "Choose your role"
                description="Use the dropdown below to switch between Member and Observer/Inactive."
                color = Color(0xF36647)
            }
            actionRow {
                stringSelect("idealync:role_select") {
                    option("Member", "member") {
                        description = "Members are also known as active contributors"
                    }
                    option("Observer/Inactive", "observer") {
                        description = "Observer/Inactive are Members who are inactive for 1+ months and are unable to contribute for now"
                    }
                }
            }
        }
    }
}